File: tests/executable/test_ir_kernel.py
"""tests/executable/test_ir_kernel.py -- unit tests for scripts/executable/ir_kernel.py and
ir_eval.py (docs/EXECUTABLE_EQUATIONS_v0_1.md sec.9, stream S2).

Covers: every row of the shared tests/executable/vectors.json (also consumed, identically, by
test_ir_kernel.js); explicit division-by-zero and out-of-domain checks (raise, never silently
return 0/inf); and a static no-float/no-mpmath guard over ir_kernel.py's and ir_eval.py's own
source, mirroring the denylist-style guard convention mcp/toledo_mcp/regex_guard.py already
uses in this repository.
"""
from __future__ import annotations

import ast
import json
import pathlib
import sys
from fractions import Fraction

import pytest

REPO_ROOT = pathlib.Path(__file__).resolve().parents[2]
EXECUTABLE_DIR = REPO_ROOT / "scripts" / "executable"

# Loaded by file path, not `from scripts.executable import ...` -- this workstation's own
# environment demonstrates a real collision (a third-party `scripts` package installed in
# site-packages shadows this repository's top-level `scripts/` directory once any sys.path
# entry ahead of the repo root supplies one), so a package-style import is not reliable here.
# ir_eval.py's own import of ir_kernel falls back to this identical sibling-file strategy for
# the same reason (see ir_eval.py's own try/except ImportError block).
sys.path.insert(0, str(EXECUTABLE_DIR))
import ir_kernel  # noqa: E402
import ir_eval  # noqa: E402

VECTORS_PATH = pathlib.Path(__file__).resolve().parent / "vectors.json"
KERNEL_SOURCE_FILES = [
    REPO_ROOT / "scripts" / "executable" / "ir_kernel.py",
    REPO_ROOT / "scripts" / "executable" / "ir_eval.py",
]


def _load_vectors() -> list[dict]:
    with open(VECTORS_PATH, "r", encoding="utf-8") as f:
        doc = json.load(f)
    assert doc["schema_version"] == "executable-test-vectors-0.1"
    return doc["vectors"]


VECTORS = _load_vectors()


def _env_from_inputs(ir: dict, inputs: dict[str, str]) -> dict[str, Fraction]:
    return {name: ir_kernel.parse_rational(value) for name, value in inputs.items()}


@pytest.mark.parametrize("vector", VECTORS, ids=[v["name"] for v in VECTORS])
def test_shared_vector_via_kernel(vector: dict) -> None:
    """Walk the vector's own 'rhs' node directly through ir_kernel.eval_node, at the term
    count its own 'transcendental' block (if any) declares, and check the result against the
    shared expected_value within the shared tolerance -- both parsed as exact Fraction, never
    float."""
    ir = vector["ir"]
    env = _env_from_inputs(ir, vector["inputs"])
    terms = ir["transcendental"]["default_terms"] if ir.get("transcendental") else ir_kernel.DEFAULT_TERMS

    result = ir_kernel.eval_node(ir["rhs"], env, terms)
    expected = ir_kernel.parse_rational(vector["expected_value"])
    tolerance = ir_kernel.parse_rational(vector["tolerance"])

    assert abs(result - expected) <= tolerance, (
        f"{vector['name']}: got {result}, expected {expected} +/- {tolerance}"
    )


@pytest.mark.parametrize("vector", VECTORS, ids=[v["name"] for v in VECTORS])
def test_shared_vector_via_ir_eval(vector: dict) -> None:
    """End-to-end: the same vector through ir_eval.evaluate (schema validation + the public
    evaluator every other stream calls), not just the bare kernel walker."""
    ir = vector["ir"]
    result = ir_eval.evaluate(ir, vector["inputs"])

    got = ir_kernel.parse_rational(result["value"])
    expected = ir_kernel.parse_rational(vector["expected_value"])
    tolerance = ir_kernel.parse_rational(vector["tolerance"])
    assert abs(got - expected) <= tolerance

    if ir.get("transcendental") is not None:
        assert result["terms_used"] == ir["transcendental"]["default_terms"]
        assert result["error_bound"] is not None
        assert ir_kernel.parse_rational(result["error_bound"]) >= 0
    else:
        assert result["terms_used"] is None
        assert result["error_bound"] is None


def test_division_by_zero_raises() -> None:
    """A 'div' node whose divisor evaluates to zero raises IRKernelError -- never silently
    returns 0 or inf (sec.9)."""
    node = {"op": "div", "args": [{"op": "const", "value": "1"}, {"op": "var", "name": "d"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="division by zero"):
        ir_kernel.eval_node(node, {"d": Fraction(0)})


def test_division_by_zero_raises_even_for_a_nonzero_declared_domain() -> None:
    """The kernel's own zero-check is unconditional -- it does not trust a variable's
    declared domain (e.g. Q_nonzero) to already exclude the value actually bound to it; the
    check re-verifies at the point of division itself."""
    node = {"op": "div", "args": [{"op": "var", "name": "a"}, {"op": "var", "name": "d"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="division by zero"):
        ir_kernel.eval_node(node, {"a": Fraction(3), "d": Fraction(0)})


def test_sqrt_of_negative_raises() -> None:
    node = {"op": "call", "fn": "sqrt", "args": [{"op": "const", "value": "-1"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="domain"):
        ir_kernel.eval_node(node, {})


def test_log_of_zero_raises() -> None:
    node = {"op": "call", "fn": "log", "args": [{"op": "const", "value": "0"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="domain"):
        ir_kernel.eval_node(node, {})


def test_log_of_negative_raises() -> None:
    node = {"op": "call", "fn": "log", "args": [{"op": "const", "value": "-5"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="domain"):
        ir_kernel.eval_node(node, {})


def test_negative_power_of_zero_raises() -> None:
    node = {"op": "pow", "args": [{"op": "const", "value": "0"}, {"op": "const", "value": "-1"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="negative power"):
        ir_kernel.eval_node(node, {})


def test_non_integer_pow_exponent_rejected() -> None:
    node = {"op": "pow", "args": [{"op": "const", "value": "2"}, {"op": "const", "value": "1/2"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="integer"):
        ir_kernel.eval_node(node, {})


def test_disallowed_operator_rejected() -> None:
    with pytest.raises(ir_kernel.IRKernelError, match="disallowed IR operator"):
        ir_kernel.eval_node({"op": "eval", "args": []}, {})


def test_disallowed_function_rejected() -> None:
    node = {"op": "call", "fn": "os.system", "args": [{"op": "const", "value": "1"}]}
    with pytest.raises(ir_kernel.IRKernelError, match="disallowed transcendental function"):
        ir_kernel.eval_node(node, {})


def test_const_node_rejects_non_string_value() -> None:
    """A 'const' node's value must be a string -- never a JSON number a lenient parser might
    otherwise silently accept and round to a float."""
    with pytest.raises(ir_kernel.IRKernelError):
        ir_kernel.eval_node({"op": "const", "value": 3}, {})  # type: ignore[dict-item]


def test_ir_eval_rejects_non_string_input() -> None:
    ir = VECTORS[0]["ir"]
    bad_inputs = dict(VECTORS[0]["inputs"])
    a_name = next(iter(bad_inputs))
    bad_inputs[a_name] = 0.5  # a float, never accepted
    with pytest.raises(ir_eval.IRValidationError):
        ir_eval.evaluate(ir, bad_inputs)


def test_ir_eval_rejects_candidate_status_transcendental_lint_mismatch() -> None:
    """sec.3/sec.13: algorithm_py and algorithm_js must never name the same family; a build-
    time lint rejects a sidecar where they match."""
    ir = json.loads(json.dumps(VECTORS[4]["ir"]))  # sqrt_two, a transcendental vector
    ir["transcendental"]["algorithm_js"] = ir["transcendental"]["algorithm_py"]
    with pytest.raises(ir_eval.IRValidationError, match="same family"):
        ir_eval.validate_ir_shape(ir)


def test_ir_eval_rejects_wrong_algorithm_py_family() -> None:
    ir = json.loads(json.dumps(VECTORS[4]["ir"]))
    ir["transcendental"]["algorithm_py"] = "not_a_real_family"
    with pytest.raises(ir_eval.IRValidationError, match="does not match"):
        ir_eval.validate_ir_shape(ir)


def test_ir_eval_requires_reviewed_by_for_reviewed_status() -> None:
    ir = json.loads(json.dumps(VECTORS[0]["ir"]))
    ir["eligibility"]["reviewed_by"] = None
    with pytest.raises(ir_eval.IRValidationError, match="reviewed_by"):
        ir_eval.validate_ir_shape(ir)


def test_ir_eval_requires_bare_output_variable_on_lhs() -> None:
    ir = json.loads(json.dumps(VECTORS[0]["ir"]))
    ir["lhs"] = {"op": "const", "value": "0"}
    with pytest.raises(ir_eval.IRValidationError, match="bare"):
        ir_eval.validate_ir_shape(ir)


def test_algorithm_families_are_pairwise_distinct_per_function() -> None:
    """ir_kernel.py's own module-level invariant, re-asserted here as a test rather than only
    an import-time assert: for every allowed transcendental fn, the declared Python-reference
    family differs from the declared JavaScript-twin family."""
    for fn in ir_kernel.ALLOWED_FNS:
        assert ir_kernel.ALGORITHM_FAMILY_PY[fn] != ir_kernel.ALGORITHM_FAMILY_JS[fn]


# ---------------------------------------------------------------------------
# Static no-float / no-mpmath guard (sec.9's own required check).
# ---------------------------------------------------------------------------

@pytest.mark.parametrize("path", KERNEL_SOURCE_FILES, ids=[p.name for p in KERNEL_SOURCE_FILES])
def test_no_float_literals_in_reference_source(path: pathlib.Path) -> None:
    """No bare Python float literal (e.g. `0.5`, `1e-9`) appears anywhere in this file's own
    AST -- the reference path is Fraction-only, always (sec.4)."""
    tree = ast.parse(path.read_text(encoding="utf-8"), filename=str(path))
    float_literals = [
        node for node in ast.walk(tree)
        if isinstance(node, ast.Constant) and isinstance(node.value, float)
    ]
    assert not float_literals, f"{path.name} contains a bare float literal: {float_literals}"


@pytest.mark.parametrize("path", KERNEL_SOURCE_FILES, ids=[p.name for p in KERNEL_SOURCE_FILES])
def test_no_mpmath_import_in_reference_source(path: pathlib.Path) -> None:
    """No `import mpmath` / `from mpmath import ...` anywhere in this file (sec.4: "mpmath
    and float never appear in this file or in ir_eval.py")."""
    tree = ast.parse(path.read_text(encoding="utf-8"), filename=str(path))
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            assert not any(alias.name.split(".")[0] == "mpmath" for alias in node.names), (
                f"{path.name} imports mpmath"
            )
        if isinstance(node, ast.ImportFrom):
            assert node.module != "mpmath" and (node.module or "").split(".")[0] != "mpmath", (
                f"{path.name} imports from mpmath"
            )


@pytest.mark.parametrize("path", KERNEL_SOURCE_FILES, ids=[p.name for p in KERNEL_SOURCE_FILES])
def test_no_bare_float_builtin_call_in_reference_source(path: pathlib.Path) -> None:
    """No call to the builtin `float(...)` anywhere in this file."""
    tree = ast.parse(path.read_text(encoding="utf-8"), filename=str(path))
    for node in ast.walk(tree):
        if isinstance(node, ast.Call) and isinstance(node.func, ast.Name) and node.func.id == "float":
            pytest.fail(f"{path.name} calls the builtin float(...)")


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(pytest.main([__file__, "-v"]))

File: tests/executable/test_ir_kernel.js
#!/usr/bin/env node
"use strict";
/*
 * tests/executable/test_ir_kernel.js -- unit tests for site/static/js/_qfrac.js and
 * _ir_eval.js (docs/EXECUTABLE_EQUATIONS_v0_1.md sec.9, stream S2), run via `node` --
 * the same runtime scripts/executable/crosscheck_runner.py (stream S3) invokes headlessly
 * against these exact shipped browser files.
 *
 * Covers: every row of the shared tests/executable/vectors.json (also consumed, identically,
 * by tests/executable/test_ir_kernel.py); explicit division-by-zero and domain checks; and a
 * grep guard asserting no native floating-point transcendental (Math.sin/Math.cos/Math.PI/
 * Math.sqrt/Math.exp/Math.log) appears anywhere in _qfrac.js/_ir_eval.js outside
 * toDecimalString's own function body.
 *
 * No test framework dependency (no jest/mocha) -- a small hand-rolled runner, consistent with
 * this feature's own "zero dependencies" constraint (sec.5) extended to its own test suite.
 * Exit code 0 iff every check passed; a failure prints which one and exits 1.
 */

const fs = require("fs");
const path = require("path");
const assert = require("assert");

const REPO_ROOT = path.resolve(__dirname, "..", "..");
const QFRAC_PATH = path.join(REPO_ROOT, "site", "static", "js", "_qfrac.js");
const IR_EVAL_PATH = path.join(REPO_ROOT, "site", "static", "js", "_ir_eval.js");
const VECTORS_PATH = path.join(__dirname, "vectors.json");

require(QFRAC_PATH);
const E = require(IR_EVAL_PATH);

const vectorsDoc = JSON.parse(fs.readFileSync(VECTORS_PATH, "utf-8"));
assert.strictEqual(vectorsDoc.schema_version, "executable-test-vectors-0.1");
const VECTORS = vectorsDoc.vectors;

let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    passed += 1;
    console.log(`ok - ${name}`);
  } catch (err) {
    failed += 1;
    console.log(`FAIL - ${name}`);
    console.log(`  ${err.message}`);
  }
}

function envFromInputs(inputs) {
  const env = {};
  for (const [name, value] of Object.entries(inputs)) {
    env[name] = QFrac.fromString(value);
  }
  return env;
}

// ---------------------------------------------------------------------
// Shared vectors, via the AST walker directly.
// ---------------------------------------------------------------------

for (const vector of VECTORS) {
  test(`shared vector: ${vector.name}`, () => {
    const ir = vector.ir;
    const env = envFromInputs(vector.inputs);
    const terms = ir.transcendental ? ir.transcendental.default_terms : E.DEFAULT_TERMS;

    const result = E.evalNode(ir.rhs, env, terms);
    const expected = QFrac.fromString(vector.expected_value);
    const tolerance = QFrac.fromString(vector.tolerance);

    const delta = result.sub(expected).abs();
    assert.ok(
      delta.cmp(tolerance) <= 0,
      `${vector.name}: got ${result.toString()}, expected ${expected.toString()} +/- ${tolerance.toString()} (delta ${delta.toString()})`
    );
  });
}

// ---------------------------------------------------------------------
// Sanity checks mirroring the Python-side edge-case tests.
// ---------------------------------------------------------------------

test("division by zero raises", () => {
  const node = { op: "div", args: [{ op: "const", value: "1" }, { op: "var", name: "d" }] };
  assert.throws(
    () => E.evalNode(node, { d: QFrac.fromInt(0) }),
    (err) => err instanceof E.IRKernelError && /division by zero/.test(err.message)
  );
});

test("division by zero raises even when the zero comes from a variable", () => {
  const node = { op: "div", args: [{ op: "var", name: "a" }, { op: "var", name: "d" }] };
  assert.throws(
    () => E.evalNode(node, { a: QFrac.fromInt(3), d: QFrac.fromInt(0) }),
    (err) => err instanceof E.IRKernelError
  );
});

test("sqrt of a negative rational raises", () => {
  const node = { op: "call", fn: "sqrt", args: [{ op: "const", value: "-1" }] };
  assert.throws(
    () => E.evalNode(node, {}),
    (err) => err instanceof E.IRKernelError && /domain/.test(err.message)
  );
});

test("log of zero raises", () => {
  const node = { op: "call", fn: "log", args: [{ op: "const", value: "0" }] };
  assert.throws(
    () => E.evalNode(node, {}),
    (err) => err instanceof E.IRKernelError && /domain/.test(err.message)
  );
});

test("log of a negative rational raises", () => {
  const node = { op: "call", fn: "log", args: [{ op: "const", value: "-5" }] };
  assert.throws(
    () => E.evalNode(node, {}),
    (err) => err instanceof E.IRKernelError && /domain/.test(err.message)
  );
});

test("negative power of zero raises", () => {
  const node = { op: "pow", args: [{ op: "const", value: "0" }, { op: "const", value: "-1" }] };
  assert.throws(
    () => E.evalNode(node, {}),
    (err) => err instanceof E.IRKernelError && /negative power/.test(err.message)
  );
});

test("non-integer pow exponent is rejected", () => {
  const node = { op: "pow", args: [{ op: "const", value: "2" }, { op: "const", value: "1/2" }] };
  assert.throws(
    () => E.evalNode(node, {}),
    (err) => err instanceof E.IRKernelError && /integer/.test(err.message)
  );
});

test("disallowed operator is rejected", () => {
  assert.throws(
    () => E.evalNode({ op: "eval", args: [] }, {}),
    (err) => err instanceof E.IRKernelError && /disallowed IR operator/.test(err.message)
  );
});

test("disallowed function is rejected", () => {
  const node = { op: "call", fn: "eval", args: [{ op: "const", value: "1" }] };
  assert.throws(
    () => E.evalNode(node, {}),
    (err) => err instanceof E.IRKernelError && /disallowed transcendental function/.test(err.message)
  );
});

// ---------------------------------------------------------------------
// R2-1 integration fix (2026-09-08): QFrac.roundToPrecision must round a negative exact tie the
// SAME way scripts/executable/ir_kernel.py::_round_to_precision does (Python's divmod-based
// rule: floor division, remainder always in [0, denominator), every exact tie toward
// +infinity) -- these are the exact hand-verified examples the finding named. Before this fix
// QFrac.roundToPrecision(-3/2, 0) returned -2 (away from zero); the Python reference returns -1.
// ---------------------------------------------------------------------

const ROUND_TO_PRECISION_CASES = [
  { num: -3n, den: 2n, digits: 0, expectedNum: -1n, expectedDen: 1n },  // -1.5 exact tie -> -1
  { num: -5n, den: 2n, digits: 0, expectedNum: -2n, expectedDen: 1n },  // -2.5 exact tie -> -2
  { num: -1n, den: 2n, digits: 0, expectedNum: 0n, expectedDen: 1n },   // -0.5 exact tie -> 0
  { num: 3n, den: 2n, digits: 0, expectedNum: 2n, expectedDen: 1n },    // +1.5 exact tie -> 2 (unaffected side)
  { num: -7n, den: 4n, digits: 1, expectedNum: -17n, expectedDen: 10n }, // -1.75 exact tie at 1 digit -> -1.7
];

for (const c of ROUND_TO_PRECISION_CASES) {
  test(`roundToPrecision negative-tie case ${c.num}/${c.den} @ ${c.digits} digits`, () => {
    const result = new QFrac(c.num, c.den).roundToPrecision(c.digits);
    const expected = new QFrac(c.expectedNum, c.expectedDen);
    assert.ok(
      result.eq(expected),
      `${c.num}/${c.den} rounded to ${c.digits} digits: got ${result.toString()}, expected ${expected.toString()}`
    );
  });
}

test("QFrac.fromString rejects a malformed literal", () => {
  assert.throws(() => QFrac.fromString("not-a-number"), (err) => err instanceof QFracError);
});

test("algorithm families are pairwise distinct per function (JS side's own view)", () => {
  for (const fn of E.ALLOWED_FNS) {
    // The JS-declared family for `fn` must differ from the Python reference's own declared
    // family for the same fn -- cross-checked here against the identical constant
    // scripts/executable/ir_kernel.py keeps for the Python side (ir_kernel.ALGORITHM_FAMILY_PY),
    // reproduced literally since this test file cannot import a Python module.
    const ALGORITHM_FAMILY_PY = {
      pi_const: "machin_arctan_series",
      exp: "taylor_series_halving_reduction",
      sin: "taylor_series_halving_reduction",
      cos: "taylor_series_halving_reduction",
      sqrt: "newton_raphson",
      log: "artanh_series_sqrt_reduction",
    };
    assert.notStrictEqual(E.ALGORITHM_FAMILY_JS[fn], ALGORITHM_FAMILY_PY[fn], `family collision for ${fn}`);
  }
});

// ---------------------------------------------------------------------
// Grep guard: no native floating-point transcendental outside toDecimalString.
// ---------------------------------------------------------------------

const FORBIDDEN_TOKENS = ["Math.sin", "Math.cos", "Math.PI", "Math.sqrt", "Math.exp", "Math.log"];

function extractFunctionBody(source, functionName) {
  const marker = `${functionName}(`;
  const start = source.indexOf(marker);
  assert.ok(start >= 0, `could not locate ${functionName}( in source`);
  const braceStart = source.indexOf("{", start);
  let depth = 0;
  let i = braceStart;
  for (; i < source.length; i++) {
    if (source[i] === "{") depth += 1;
    else if (source[i] === "}") {
      depth -= 1;
      if (depth === 0) break;
    }
  }
  return source.slice(braceStart, i + 1);
}

test("no native Math transcendental outside toDecimalString in _qfrac.js/_ir_eval.js", () => {
  for (const filePath of [QFRAC_PATH, IR_EVAL_PATH]) {
    const source = fs.readFileSync(filePath, "utf-8");
    let sourceOutsideAllowedException = source;
    if (source.includes("toDecimalString(")) {
      const body = extractFunctionBody(source, "toDecimalString");
      sourceOutsideAllowedException = source.replace(body, "");
    }
    for (const token of FORBIDDEN_TOKENS) {
      assert.ok(
        !sourceOutsideAllowedException.includes(token),
        `${path.basename(filePath)} contains ${token} outside toDecimalString's own function body`
      );
    }
  }
});

test("toDecimalString itself uses no native Math transcendental either (BigInt long division only)", () => {
  const source = fs.readFileSync(QFRAC_PATH, "utf-8");
  const body = extractFunctionBody(source, "toDecimalString");
  for (const token of FORBIDDEN_TOKENS) {
    assert.ok(!body.includes(token), `toDecimalString itself contains ${token}`);
  }
});

// ---------------------------------------------------------------------

console.log(`\n${passed} passed, ${failed} failed`);
process.exit(failed === 0 ? 0 : 1);
